stored_file.php

```diff
@@ -0,0 +1,387 @@
+
+        if ($supported_exts && is_array($supported_exts) && count($supported_exts))
+        {
+            return in_array($this->extension,
+                $supported_exts);
+        }
+
+        return in_array($this->extension,
+            static::$__supported_exts);
+    }
+
+    public function exists()
+    {
+        return is_file($this->actual_filename());
+    }
+
+    public function size()
+    {
+        return filesize($this->actual_filename());
+    }
+
+    public function human_size()
+    {
+        $size = $this->size();
+
+        if ($size > static::SIZE_GIB)
+        {
+            $size = ($size / static::SIZE_GIB);
+            return sprintf('%.2f GiB', $size);
+        }
+
+        if ($size > static::SIZE_MIB)
+        {
+            $size = ($size / static::SIZE_MIB);
+            return sprintf('%.2f MiB', $size);
+        }
+
+        $size = ($size / static::SIZE_KIB);
+        return sprintf('%.2f KiB', $size);
+    }
+
+    public function detect_mime()
+    {
+        if ($this->mime) return $this->mime;
+        $this->mime = File_Util::detect_mime($this->actual_filename());
+        return $this->mime;
+    }
+
+    public function generate_filename($ext = null)
+    {
+        if (!$ext) $ext = $this->extension;
+
+        $hash = md5_file($this->source);
+        $dir1 = substr($hash, 0, 2);
+        $dir2 = substr($hash, 2, 2);
+        $name = substr($hash, 4, 28);
+
+        $this->filename = build_path($dir1, $dir2,
+            sf('%s.%s', $name, $ext));
+        return $this->filename;
+    }
+
+    public function read()
+    {
+        return file_get_contents($this->actual_filename());
+    }
+
+    public function move($invoker_tag = null)
+    {
+        if ($this->__moved) return;
+
+        if (!$invoker_tag) $invoker_tag = md5(serialize(debug_backtrace(5)));
+
+        $destination = static::resolve_file($this->filename, true);
+        $dir = dirname($destination);
+        $ref_filename = $destination . '.ref';
+
+        if (!is_dir($dir)) mkdir($dir, 0755, true);
+
+        if (!file_exists($destination)) {
+
+            rename($this->source, $destination);
+
+            // Create a new file reference
+            $ref_res = fopen($ref_filename, 'w');
+            if (flock($ref_res, LOCK_EX)) {
+                ftruncate($ref_res, 0);
+                fwrite($ref_res, "{\"ref_count\": 1,\"atags\" : [\"$invoker_tag\"],\"dtags\" : []}");
+                fflush($ref_res);
+                flock($ref_res, LOCK_UN);
+                fclose($ref_res);
+            } else {
+                throw new \RuntimeException("Couldn't get the lock!");
+            }
+
+        } elseif (file_exists($ref_filename)) {
+
+            // Update the existent file reference according to invoker_id
+            $ref_res = fopen($ref_filename, 'r+');
+            if (flock($ref_res, LOCK_EX)) {
+
+                $ref_content = json_decode(fread($ref_res, filesize($ref_filename)), true);
```

Comment:
No attempt at detecting the `FALSE` on failure value. If the network-based file system fails momentarily you could be using `FALSE` and then still attempting to manipulate it later on. Possible loss of data.

Reply:
I did not apply extensive checks because the current class does not do that too, so I just expand the basic functionality as you asked me. I can easily improve code as requested, just tell me where and when.